coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt

```diff
@@ -0,0 +1,263 @@
+    loading: @Composable (() -> Unit)? = null,
+    modifier: Modifier = Modifier,
+    onRequestCompleted: (RequestResult) -> Unit = emptySuccessLambda
+) {
+    CoilImage(
+        request = request,
+        alignment = alignment,
+        contentScale = contentScale,
+        getSuccessPainter = { crossfadePainter(it, durationMs = crossfadeDuration) },
+        getFailurePainter = getFailurePainter,
+        loading = loading,
+        modifier = modifier,
+        onRequestCompleted = onRequestCompleted
+    )
+}
+
+/**
+ * A composable function which runs an fade animation on the given [result], returning the
+ * [Painter] which should be used to paint the [ImageAsset].
+ *
+ * The animation fades in the image's saturation, alpha and exposure. More information on the
+ * pattern can be seen [here](https://material.io/archive/guidelines/patterns/loading-images.html).
+ *
+ * @param result The result of a image fetch.
+ * @param skipFadeWhenLoadedFromMemory Whether the fade animation should be skipped when the result
+ * has been loaded from memory.
+ * @param durationMs The duration of the crossfade animation in milliseconds.
+ * @param clock The animation clock.
+ */
+@Composable
+private fun crossfadePainter(
+    result: SuccessResult,
+    skipFadeWhenLoadedFromMemory: Boolean = true,
+    durationMs: Int = DefaultTransitionDuration,
+    clock: AnimationClockObservable = AnimationClockAmbient.current.asDisposableClock()
+): Painter {
+    return if (skipFadeWhenLoadedFromMemory && result.isFromMemory) {
+        // If can skip the fade when loaded from memory, we do not need to run the animation on it
+        defaultSuccessPainterGetter(result)
+    } else {
+        val observablePainter = remember {
+            ObservableCrossfade(result, durationMs, clock).also { it.start() }
+        }
+        if (!observablePainter.isFinished) {
+            // If the animation is running, using it's painter
+            observablePainter.painter
+        } else {
+            // If the animation has finished, revert back to the default painter
+            defaultSuccessPainterGetter(result)
+        }
+    }
+}
+
+@Suppress("JoinDeclarationAndAssignment")
+private class ObservableCrossfade(
+    result: SuccessResult,
+    crossfadeDuration: Int,
+    clock: AnimationClockObservable
+) {
+    var painter: Painter by mutableStateOf(
+        createCrossfadePainter(
+            image = result.image,
+            saturation = 0f,
+            alpha = 0f,
+            brightness = 0f
+        )
+    )
+
+    var isFinished by mutableStateOf(false)
+        private set
+
+    private val animation: TransitionAnimation<CrossfadeTransition.State>
+
+    init {
+        animation = CrossfadeTransition.definition(crossfadeDuration).createAnimation(clock)
+
+        animation.onUpdate = {
+            // Animation tick, so update the painter using the current transition property values
+            painter = createCrossfadePainter(
```

Comment:
creating this on every tick seems expensive. You can reuse a ImageLoadingColorMatrix in createCrossfadePainter. Can you mutate ColorMatrixImagePainter?

Reply:
Yeah, fixed this. Had to collapse down some classes, but far less objects are created now.